File: Binary Tree/235. Lowest Common Ancestor of a Binary Search Tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
// Time complexity is O(n) space complexity O(n)

struct TreeNode *lowestCommonAncestor(struct TreeNode *root, struct TreeNode *p, struct TreeNode *q)
{
    if (root == NULL)
    {
        return NULL;
    }

    struct TreeNode *left = lowestCommonAncestor(root->left, p, q);
    struct TreeNode *right = lowestCommonAncestor(root->right, p, q);

    if (root == p || root == q)
    {
        return root;
    }

    if (left != NULL && right != NULL)
    {
        return root;
    }
    else if (left != NULL && right == NULL)
    {
        return left;
    }
    else if (right != NULL && left == NULL)
    {
        return right;
    }

    return NULL;
}

File: Binary Search Tree/235. Lowest Common Ancestor of a Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        currentPtr = root
        
        while currentPtr is not None:
            if p.val > currentPtr.val and q.val > currentPtr.val:
                currentPtr = currentPtr.right
            elif p.val < currentPtr.val and q.val < currentPtr.val:
                currentPtr = currentPtr.left
            else:
                break
                
                
        return currentPtr